src/hotspot/share/classfile/classLoaderDataShared.cpp

```diff
@@ -0,0 +1,202 @@
+  assert(UseSharedSpaces, "must be");
+  assert_valid(loader_data);
+  if (_modules != NULL) { // Could be NULL if we have archived no modules for platform/system loaders
+    ModuleEntryTable* modules = loader_data->modules();
+    PackageEntryTable* packages = loader_data->packages();
+
+    MutexLocker m1(Module_lock);
+    if (do_entries) {
+      modules->load_archived_entries(loader_data, _modules);
+      packages->load_archived_entries(_packages);
+    }
+    if (do_oops) {
+      modules->restore_archived_oops(loader_data, _modules);
+    }
+  }
+}
+
+// ------------------------------
+
+ClassLoaderData* ClassLoaderDataShared::null_class_loader_data() {
+  ClassLoaderData* loader_data = ClassLoaderData::the_null_class_loader_data();
+  assert(loader_data != NULL, "must be");
+  return loader_data;
+}
+
+ClassLoaderData* ClassLoaderDataShared::java_platform_loader_data_or_null() {
+  return ClassLoaderData::class_loader_data_or_null(SystemDictionary::java_platform_loader());
+}
+
+ClassLoaderData* ClassLoaderDataShared::java_system_loader_data_or_null() {
+  return ClassLoaderData::class_loader_data_or_null(SystemDictionary::java_system_loader());
+}
+
+void ClassLoaderDataShared::iterate_symbols(MetaspaceClosure* closure) {
+  assert(DumpSharedSpaces && MetaspaceShared::use_full_module_graph(), "must be");
+  _archived_boot_loader_data.iterate_symbols    (null_class_loader_data(), closure);
+  _archived_platform_loader_data.iterate_symbols(java_platform_loader_data_or_null(), closure);
+  _archived_system_loader_data.iterate_symbols  (java_system_loader_data_or_null(), closure);
+}
+
+void ClassLoaderDataShared::allocate_archived_tables() {
+  assert(DumpSharedSpaces && MetaspaceShared::use_full_module_graph(), "must be");
+  _archived_boot_loader_data.allocate    (null_class_loader_data());
+  _archived_platform_loader_data.allocate(java_platform_loader_data_or_null());
+  _archived_system_loader_data.allocate  (java_system_loader_data_or_null());
+}
+
+void ClassLoaderDataShared::init_archived_tables() {
+  assert(DumpSharedSpaces && MetaspaceShared::use_full_module_graph(), "must be");
+  _archived_boot_loader_data.init_archived_entries    (null_class_loader_data());
+  _archived_platform_loader_data.init_archived_entries(java_platform_loader_data_or_null());
+  _archived_system_loader_data.init_archived_entries  (java_system_loader_data_or_null());
+  _archived_javabase_moduleEntry = ModuleEntry::get_archived_entry(ModuleEntryTable::javabase_moduleEntry());
+}
+
+void ClassLoaderDataShared::init_archived_oops() {
+  assert(DumpSharedSpaces && MetaspaceShared::use_full_module_graph(), "must be");
+  _archived_boot_loader_data.init_archived_oops    (null_class_loader_data());
+  _archived_platform_loader_data.init_archived_oops(java_platform_loader_data_or_null());
+  _archived_system_loader_data.init_archived_oops  (java_system_loader_data_or_null());
+}
+
+void ClassLoaderDataShared::serialize(class SerializeClosure* f) {
```

Comment:
I fixed this one and other issues you reported for the same commit (I didn't want to respond to each one individually to avoid the e-mail avalanche). New version: [e541890](https://github.com/openjdk/jdk/pull/80/commits/e541890e037ff40ec9134a54e5c7a878ab9259f3)